File: iosApp/iosApp/Views/TaskCompletionBar/TaskCompletionBarViewModel.swift
//
//  TaskCompletionBarViewModel.swift
//  More
//
//  Created by Julia Mayrhauser on 26.04.23.
//  Copyright © 2023 Ludwig Boltzmann Institute for
//  Digital Health and Prevention - A research institute
//  of the Ludwig Boltzmann Gesellschaft,
//  Oesterreichische Vereinigung zur Foerderung
//  der wissenschaftlichen Forschung
//  Licensed under the Apache 2.0 license with Commons Clause
//  (see https://www.apache.org/licenses/LICENSE-2.0 and
//  https://commonsclause.com/).
//

import Combine
import KMPNativeCoroutinesCombine
import shared

class TaskCompletionBarViewModel: ObservableObject {
    @Published var taskCompletion: TaskCompletion = TaskCompletion(finishedTasks: 0, totalTasks: 0)
    @Published var taskCompletionPercentage: Double = 0
    var coreViewModel = CoreTaskCompletionBarViewModel(repository: AppDelegate.shared.repositories, dispatcher: AppDispatchers.shared.default_)

    private var cancellables: Set<AnyCancellable> = []

    init() {
        createPublisher(for: coreViewModel.taskCompletion)
        .receive(on: DispatchQueue.main)
        .sink(receiveCompletion: { _ in }) { [weak self] completion in
            self?.taskCompletion = completion
            self?.taskCompletionPercentage = (Double(completion.finishedTasks) / Double(completion.totalTasks)) * 100
        }
        .store(in: &cancellables)
    }
}

File: androidApp/src/main/java/io/redlink/umm/blendedcare/app/android/activities/taskCompletion/TaskCompletionBarViewModel.kt
/*
 * Copyright LBI-DHP and/or licensed to LBI-DHP under one or more
 * contributor license agreements (LBI-DHP: Ludwig Boltzmann Institute
 * for Digital Health and Prevention -- A research institute of the
 * Ludwig Boltzmann Gesellschaft, Österreichische Vereinigung zur
 * Förderung der wissenschaftlichen Forschung).
 * Licensed under the Apache 2.0 license with Commons Clause
 * (see https://www.apache.org/licenses/LICENSE-2.0 and
 * https://commonsclause.com/).
 */
package io.redlink.umm.blendedcare.app.android.activities.taskCompletion

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.redlink.umm.blendedcare.app.android.BlendedCareApplication
import io.redlink.umm.participant.models.TaskCompletion
import io.redlink.umm.participant.viewModels.taskCompletionBar.CoreTaskCompletionBarViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class TaskCompletionBarViewModel : ViewModel() {
    private val coreViewModel =
        CoreTaskCompletionBarViewModel(BlendedCareApplication.Companion.shared!!.repositories)
    val taskCompletion: MutableState<TaskCompletion> = mutableStateOf(TaskCompletion())

    init {
        viewModelScope.launch(Dispatchers.IO) {
            coreViewModel.taskCompletion.collect {
                withContext(Dispatchers.Main) {
                    taskCompletion.value = it
                }
            }
        }
    }
}